planner/core/logical_plan_builder.go

```diff
@@ -5111,10 +5111,10 @@ type windowFuncs struct {
 
 // sortWindowSpecs sorts the window specifications by reversed alphabetical order, then we could add less `Sort` operator
 // in physical plan because the window functions with the same partition by and order by clause will be at near places.
-func sortWindowSpecs(groupedFuncs map[*ast.WindowSpec][]*ast.WindowFuncExpr) []windowFuncs {
+func sortWindowSpecs(groupedFuncs map[*ast.WindowSpec][]*ast.WindowFuncExpr, orderedSpec []*ast.WindowSpec) []windowFuncs {
 	windows := make([]windowFuncs, 0, len(groupedFuncs))
-	for spec, funcs := range groupedFuncs {
-		windows = append(windows, windowFuncs{spec, funcs})
+	for _, spec := range orderedSpec {
+		windows = append(windows, windowFuncs{spec, groupedFuncs[spec]})
```

Comment:
Can we just modify https://github.com/pingcap/tidb/pull/22853/files#diff-3b47fd3bf2c33ae3a03037a65ed39df33a4ae73debe9cf6f5bb2f397508061a0R5121 to make this order stable?

Reply:
> Can we just modify https://github.com/pingcap/tidb/pull/22853/files#diff-3b47fd3bf2c33ae3a03037a65ed39df33a4ae73debe9cf6f5bb2f397508061a0R5121 to make this order stable?

@eurekaka The previous author wanted to use `sort.SliceStable` to sort spec stably. If 2 specs are the same, the order is the only which relies on input order of the 2. Unfortunately, the input order cannot be guaranteed because of the `map` implementation in `Golang`.

Reply:
@eurekaka U can check unit test case below for more details.